### PR TITLE
For issue #2114: Validation tests

### PR DESCRIPTION
**Pull request by a-saf**
This PR is for issue #2114

### CODE REVIEW

File: dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    @Test
+    public void testCorrectLinkWithRef() {
+        LinkValidator validator = LinkValidator.getInstance();
+        TestValidationHelper vh = new TestValidationHelper();
+
+        LinkImpl correctLink = new LinkImpl();
+        correctLink.setDescription("This is a correct test link");
+        String operationRef = "#/paths/my-test-path-two/get";
```
Comment (leochr) [inline dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java]: this should be:
String operationRef = "#/paths/~1my-test-path-two/get";

The slashes in path should be escaped with ~1

File: dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/impl/validation/LinkValidator.java [hunk 2]
```diff
@@ -69,7 +69,7 @@ public void validate(ValidationHelper helper, Context context, String key, Link
                         if (operationRef.length != 4) {
                             isValid = false;
                         } else {
-                            String pathKey = operationRef[2].replace("~1", "/").replace("~0", "~");
+                            String pathKey = "/" + operationRef[2].replace("~1", "/").replace("~0", "~");
```
Comment (leochr) [inline dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/impl/validation/LinkValidator.java]: All slashes in the path should be escaped with ~1. A slash shouldn't be added in-front. The test needs to be corrected to have ~1 (see comments in LinkValidatorTest)

Comment (leochr) [reply dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/impl/validation/LinkValidator.java]: Also noticed that the checks above wouldn't catch invalid cases such as `#/invalid/~1myapp~1repositories/get`

So add a check like this:

change this:
```
String[] operationRef = t.getOperationRef().split("/");
if (operationRef.length != 4) {
   isValid = false;
```
to:

```
String[] operationRef;
if (!t.getOperationRef().startsWith("#/paths/") || (operationRef = t.getOperationRef().split("/")).length != 4) {
    isValid = false;
```

File: dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    public void testLinkWithBothRefAndId() {
+        LinkValidator validator = LinkValidator.getInstance();
+        TestValidationHelper vh = new TestValidationHelper();
+
+        LinkImpl link = new LinkImpl();
+        link.setDescription("This is a link with both OperationRef and operationId defined");
+        String operationId = "pathItemOneGetId";
+        link.setOperationId(operationId);
+
+        String operationRef = "#/paths/my-test-path-two/get";
```
Comment (leochr) [inline dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java]: Need to change to
`String operationRef = "#/paths/~1my-test-path-two/get";`

File: dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    @Test
+    public void testLinkWithInvalidOperation() {
+        LinkValidator validator = LinkValidator.getInstance();
+        TestValidationHelper vh = new TestValidationHelper();
+
+        LinkImpl correctLink = new LinkImpl();
+        correctLink.setDescription("This is a link with invalid operation in operationRef");
+        String operationRef = "#/paths/my-test-path-two/write";
```
Comment (leochr) [inline dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java]: Add ~1

File: dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        correctLink.setOperationRef(operationRef);
+
+        validator.validate(vh, context, key, correctLink);
+        Assert.assertEquals(1, vh.getEventsSize());
+        Assert.assertTrue(vh.getResult().getEvents().get(0).message.contains("Link Object specified a relative \"operationRef\" field"));
+    }
+
+    @Test
+    public void testLinkWithCorrectId() {
```
Comment (leochr) [inline dev/com.ibm.ws.microprofile.openapi_test/test/com/ibm/ws/microprofile/openapi/validation/test/LinkValidatorTest.java]: seems this test is same as testCorrectLinkWithId?

**Review by leochr — COMMENTED** [review]
@a-saf looks good. some comments. Thanks!

**Comment by a-saf** [comment]
@leochr Fixed!

**Review by leochr — APPROVED** [review]
thanks @a-saf!

**Pull request merged**